File: frontend/src/services/audit.service.ts
import api from './api.service';

export interface AuditLog {
  id: number;
  timestamp: string;
  userId?: number;
  userName?: string;
  action: string;
  entityName: string;
  entityId?: string;
  oldValues?: string;
  newValues?: string;
  details?: string;
  ipAddress?: string;
  userAgent?: string;
  storeId?: number;
  storeName?: string;
}

export interface SecurityLog {
  id: number;
  timestamp: string;
  eventType: number;
  eventTypeName: string;
  severity: number;
  severityName: string;
  userId?: number;
  userName?: string;
  description: string;
  ipAddress?: string;
  userAgent?: string;
  success: boolean;
  metadata?: string;
  storeId?: number;
  storeName?: string;
}

export interface AuditLogSearchRequest {
  startDate?: string;
  endDate?: string;
  userId?: number;
  action?: string;
  entityName?: string;
  storeId?: number;
  page?: number;
  pageSize?: number;
}

export interface SecurityLogSearchRequest {
  startDate?: string;
  endDate?: string;
  userId?: number;
  eventType?: number;
  severity?: number;
  success?: boolean;
  storeId?: number;
  page?: number;
  pageSize?: number;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
  hasPrevious: boolean;
  hasNext: boolean;
}

export interface AuditStatistics {
  totalAuditLogs: number;
  totalSecurityLogs: number;
  failedLoginAttempts24h: number;
  unauthorizedAccess24h: number;
  criticalEvents7d: number;
  topUsers: TopUserActivity[];
  activityByDay: ActivityByDay[];
  securityEventCounts: SecurityEventCount[];
}

export interface TopUserActivity {
  userId: number;
  userName: string;
  activityCount: number;
}

export interface ActivityByDay {
  date: string;
  count: number;
}

export interface SecurityEventCount {
  eventType: number;
  eventTypeName: string;
  count: number;
}

const auditService = {
  // Audit Logs
  getAuditLogs: async (request: AuditLogSearchRequest): Promise<PagedResult<AuditLog>> => {
    const params = new URLSearchParams();
    if (request.startDate) params.append('startDate', request.startDate);
    if (request.endDate) params.append('endDate', request.endDate);
    if (request.userId) params.append('userId', request.userId.toString());
    if (request.action) params.append('action', request.action);
    if (request.entityName) params.append('entityName', request.entityName);
    if (request.storeId) params.append('storeId', request.storeId.toString());
    if (request.page) params.append('page', request.page.toString());
    if (request.pageSize) params.append('pageSize', request.pageSize.toString());

    const response = await api.get(`/audit/audit-logs?${params.toString()}`);
    return response.data.data;
  },

  getAuditLogById: async (id: number): Promise<AuditLog> => {
    const response = await api.get(`/audit/audit-logs/${id}`);
    return response.data.data;
  },

  // Security Logs
  getSecurityLogs: async (request: SecurityLogSearchRequest): Promise<PagedResult<SecurityLog>> => {
    const params = new URLSearchParams();
    if (request.startDate) params.append('startDate', request.startDate);
    if (request.endDate) params.append('endDate', request.endDate);
    if (request.userId) params.append('userId', request.userId.toString());
    if (request.eventType !== undefined) params.append('eventType', request.eventType.toString());
    if (request.severity !== undefined) params.append('severity', request.severity.toString());
    if (request.success !== undefined) params.append('success', request.success.toString());
    if (request.storeId) params.append('storeId', request.storeId.toString());
    if (request.page) params.append('page', request.page.toString());
    if (request.pageSize) params.append('pageSize', request.pageSize.toString());

    const response = await api.get(`/audit/security-logs?${params.toString()}`);
    return response.data.data;
  },

  getSecurityLogById: async (id: number): Promise<SecurityLog> => {
    const response = await api.get(`/audit/security-logs/${id}`);
    return response.data.data;
  },

  // Statistics
  getStatistics: async (days: number = 30): Promise<AuditStatistics> => {
    const response = await api.get(`/audit/statistics?days=${days}`);
    return response.data.data;
  },

  getRecentActivity: async (count: number = 20): Promise<AuditLog[]> => {
    const response = await api.get(`/audit/recent-activity?count=${count}`);
    return response.data.data;
  },

  getRecentSecurityEvents: async (count: number = 20): Promise<SecurityLog[]> => {
    const response = await api.get(`/audit/recent-security-events?count=${count}`);
    return response.data.data;
  },

  // Export
  exportAuditLogs: async (request: AuditLogSearchRequest): Promise<Blob> => {
    const response = await api.post('/audit/audit-logs/export', request, {
      responseType: 'blob'
    });
    return response.data;
  },

  exportSecurityLogs: async (request: SecurityLogSearchRequest): Promise<Blob> => {
    const response = await api.post('/audit/security-logs/export', request, {
      responseType: 'blob'
    });
    return response.data;
  },

  // Enums
  getSecurityEventTypes: async (): Promise<Record<number, string>> => {
    const response = await api.get('/audit/security-event-types');
    return response.data.data;
  },

  getSecuritySeverityLevels: async (): Promise<Record<number, string>> => {
    const response = await api.get('/audit/security-severity-levels');
    return response.data.data;
  }
};

export default auditService;

File: backend/src/POS.Infrastructure/Services/AuditService.cs
using Microsoft.EntityFrameworkCore;
using POS.Application.DTOs.Audit;
using POS.Application.Interfaces;
using POS.Domain.Entities.Audit;
using POS.Infrastructure.Data;
using System.Text;

namespace POS.Infrastructure.Services;

public class AuditService : IAuditService
{
    private readonly POSDbContext _context;

    public AuditService(POSDbContext context)
    {
        _context = context;
    }

    #region Audit Logging

    public async Task LogAuditAsync(AuditLog auditLog, CancellationToken cancellationToken = default)
    {
        await _context.AuditLogs.AddAsync(auditLog, cancellationToken);
        await _context.SaveChangesAsync(cancellationToken);
    }

    public async Task<PagedResult<AuditLogDto>> GetAuditLogsAsync(
        AuditLogSearchRequest request, 
        CancellationToken cancellationToken = default)
    {
        var query = _context.AuditLogs
            .Include(a => a.Store)
            .AsQueryable();

        // Apply filters
        if (request.StartDate.HasValue)
            query = query.Where(a => a.Timestamp >= request.StartDate.Value);

        if (request.EndDate.HasValue)
            query = query.Where(a => a.Timestamp <= request.EndDate.Value);

        if (request.UserId.HasValue)
            query = query.Where(a => a.UserId == request.UserId.Value);

        if (!string.IsNullOrEmpty(request.Action))
            query = query.Where(a => a.Action.Contains(request.Action));

        if (!string.IsNullOrEmpty(request.EntityName))
            query = query.Where(a => a.EntityName.Contains(request.EntityName));

        if (request.StoreId.HasValue)
            query = query.Where(a => a.StoreId == request.StoreId.Value);

        // Get total count
        var totalCount = await query.CountAsync(cancellationToken);

        // Apply pagination
        var items = await query
            .OrderByDescending(a => a.Timestamp)
            .Skip((request.Page - 1) * request.PageSize)
            .Take(request.PageSize)
            .Select(a => new AuditLogDto
            {
                Id = a.Id,
                Timestamp = a.Timestamp,
                UserId = a.UserId,
                UserName = a.UserName,
                Action = a.Action,
                EntityName = a.EntityName,
                EntityId = a.EntityId,
                OldValues = a.OldValues,
                NewValues = a.NewValues,
                Details = a.Details,
                IpAddress = a.IpAddress,
                UserAgent = a.UserAgent,
                StoreId = a.StoreId,
                StoreName = a.Store != null ? a.Store.Name : null
            })
            .ToListAsync(cancellationToken);

        return new PagedResult<AuditLogDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = request.Page,
            PageSize = request.PageSize
        };
    }

    public async Task<AuditLogDto?> GetAuditLogByIdAsync(long id, CancellationToken cancellationToken = default)
    {
        return await _context.AuditLogs
            .Include(a => a.Store)
            .Where(a => a.Id == id)
            .Select(a => new AuditLogDto
            {
                Id = a.Id,
                Timestamp = a.Timestamp,
                UserId = a.UserId,
                UserName = a.UserName,
                Action = a.Action,
                EntityName = a.EntityName,
                EntityId = a.EntityId,
                OldValues = a.OldValues,
                NewValues = a.NewValues,
                Details = a.Details,
                IpAddress = a.IpAddress,
                UserAgent = a.UserAgent,
                StoreId = a.StoreId,
                StoreName = a.Store != null ? a.Store.Name : null
            })
            .FirstOrDefaultAsync(cancellationToken);
    }

    #endregion

    #region Security Logging

    public async Task LogSecurityEventAsync(SecurityLog securityLog, CancellationToken cancellationToken = default)
    {
        await _context.SecurityLogs.AddAsync(securityLog, cancellationToken);
        await _context.SaveChangesAsync(cancellationToken);
    }

    public async Task<PagedResult<SecurityLogDto>> GetSecurityLogsAsync(
        SecurityLogSearchRequest request, 
        CancellationToken cancellationToken = default)
    {
        var query = _context.SecurityLogs
            .Include(s => s.Store)
            .AsQueryable();

        // Apply filters
        if (request.StartDate.HasValue)
            query = query.Where(s => s.Timestamp >= request.StartDate.Value);

        if (request.EndDate.HasValue)
            query = query.Where(s => s.Timestamp <= request.EndDate.Value);

        if (request.UserId.HasValue)
            query = query.Where(s => s.UserId == request.UserId.Value);

        if (request.EventType.HasValue)
            query = query.Where(s => s.EventType == request.EventType.Value);

        if (request.Severity.HasValue)
            query = query.Where(s => s.Severity == request.Severity.Value);

        if (request.Success.HasValue)
            query = query.Where(s => s.Success == request.Success.Value);

        if (request.StoreId.HasValue)
            query = query.Where(s => s.StoreId == request.StoreId.Value);

        // Get total count
        var totalCount = await query.CountAsync(cancellationToken);

        // Apply pagination
        var items = await query
            .OrderByDescending(s => s.Timestamp)
            .Skip((request.Page - 1) * request.PageSize)
            .Take(request.PageSize)
            .Select(s => new SecurityLogDto
            {
                Id = s.Id,
                Timestamp = s.Timestamp,
                EventType = s.EventType,
                EventTypeName = s.EventType.ToString(),
                Severity = s.Severity,
                SeverityName = s.Severity.ToString(),
                UserId = s.UserId,
                UserName = s.UserName,
                Description = s.Description,
                IpAddress = s.IpAddress,
                UserAgent = s.UserAgent,
                Success = s.Success,
                Metadata = s.Metadata,
                StoreId = s.StoreId,
                StoreName = s.Store != null ? s.Store.Name : null
            })
            .ToListAsync(cancellationToken);

        return new PagedResult<SecurityLogDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = request.Page,
            PageSize = request.PageSize
        };
    }

    public async Task<SecurityLogDto?> GetSecurityLogByIdAsync(long id, CancellationToken cancellationToken = default)
    {
        return await _context.SecurityLogs
            .Include(s => s.Store)
            .Where(s => s.Id == id)
            .Select(s => new SecurityLogDto
            {
                Id = s.Id,
                Timestamp = s.Timestamp,
                EventType = s.EventType,
                EventTypeName = s.EventType.ToString(),
                Severity = s.Severity,
                SeverityName = s.Severity.ToString(),
                UserId = s.UserId,
                UserName = s.UserName,
                Description = s.Description,
                IpAddress = s.IpAddress,
                UserAgent = s.UserAgent,
                Success = s.Success,
                Metadata = s.Metadata,
                StoreId = s.StoreId,
                StoreName = s.Store != null ? s.Store.Name : null
            })
            .FirstOrDefaultAsync(cancellationToken);
    }

    #endregion

    #region Statistics & Analytics

    public async Task<AuditStatisticsDto> GetAuditStatisticsAsync(int days = 30, CancellationToken cancellationToken = default)
    {
        var startDate = DateTime.Now.AddDays(-days);
        var last24Hours = DateTime.Now.AddHours(-24);
        var last7Days = DateTime.Now.AddDays(-7);

        var statistics = new AuditStatisticsDto
        {
            TotalAuditLogs = await _context.AuditLogs.CountAsync(cancellationToken),
            TotalSecurityLogs = await _context.SecurityLogs.CountAsync(cancellationToken),
            
            FailedLoginAttempts24h = await _context.SecurityLogs
                .Where(s => s.Timestamp >= last24Hours && 
                           s.EventType == SecurityEventType.LoginFailed)
                .CountAsync(cancellationToken),
            
            UnauthorizedAccess24h = await _context.SecurityLogs
                .Where(s => s.Timestamp >= last24Hours && 
                           s.EventType == SecurityEventType.UnauthorizedAccess)
                .CountAsync(cancellationToken),
            
            CriticalEvents7d = await _context.SecurityLogs
                .Where(s => s.Timestamp >= last7Days && 
                           s.Severity == SecuritySeverity.Critical)
                .CountAsync(cancellationToken),
            
            TopUsers = await _context.AuditLogs
                .Where(a => a.Timestamp >= startDate && a.UserId.HasValue)
                .GroupBy(a => new { a.UserId, a.UserName })
                .Select(g => new TopUserActivityDto
                {
                    UserId = g.Key.UserId!.Value,
                    UserName = g.Key.UserName ?? "Unknown",
                    ActivityCount = g.Count()
                })
                .OrderByDescending(u => u.ActivityCount)
                .Take(10)
                .ToListAsync(cancellationToken),
            
            ActivityByDay = await _context.AuditLogs
                .Where(a => a.Timestamp >= startDate)
                .GroupBy(a => a.Timestamp.Date)
                .Select(g => new ActivityByDayDto
                {
                    Date = g.Key,
                    Count = g.Count()
                })
                .OrderBy(a => a.Date)
                .ToListAsync(cancellationToken),
            
            SecurityEventCounts = await _context.SecurityLogs
                .Where(s => s.Timestamp >= startDate)
                .GroupBy(s => s.EventType)
                .Select(g => new SecurityEventCountDto
                {
                    EventType = g.Key,
                    EventTypeName = g.Key.ToString(),
                    Count = g.Count()
                })
                .OrderByDescending(e => e.Count)
                .ToListAsync(cancellationToken)
        };

        return statistics;
    }

    public async Task<List<AuditLogDto>> GetRecentActivityAsync(int count = 20, CancellationToken cancellationToken = default)
    {
        return await _context.AuditLogs
            .Include(a => a.Store)
            .OrderByDescending(a => a.Timestamp)
            .Take(count)
            .Select(a => new AuditLogDto
            {
                Id = a.Id,
                Timestamp = a.Timestamp,
                UserId = a.UserId,
                UserName = a.UserName,
                Action = a.Action,
                EntityName = a.EntityName,
                EntityId = a.EntityId,
                Details = a.Details,
                IpAddress = a.IpAddress,
                StoreId = a.StoreId,
                StoreName = a.Store != null ? a.Store.Name : null
            })
            .ToListAsync(cancellationToken);
    }

    public async Task<List<SecurityLogDto>> GetRecentSecurityEventsAsync(int count = 20, CancellationToken cancellationToken = default)
    {
        return await _context.SecurityLogs
            .Include(s => s.Store)
            .OrderByDescending(s => s.Timestamp)
            .Take(count)
            .Select(s => new SecurityLogDto
            {
                Id = s.Id,
                Timestamp = s.Timestamp,
                EventType = s.EventType,
                EventTypeName = s.EventType.ToString(),
                Severity = s.Severity,
                SeverityName = s.Severity.ToString(),
                UserId = s.UserId,
                UserName = s.UserName,
                Description = s.Description,
                IpAddress = s.IpAddress,
                Success = s.Success,
                StoreId = s.StoreId,
                StoreName = s.Store != null ? s.Store.Name : null
            })
            .ToListAsync(cancellationToken);
    }

    #endregion

    #region Export

    public async Task<byte[]> ExportAuditLogsAsync(AuditLogSearchRequest request, CancellationToken cancellationToken = default)
    {
        var logs = await GetAuditLogsAsync(new AuditLogSearchRequest
        {
            StartDate = request.StartDate,
            EndDate = request.EndDate,
            UserId = request.UserId,
            Action = request.Action,
            EntityName = request.EntityName,
            StoreId = request.StoreId,
            Page = 1,
            PageSize = 10000 // Large page for export
        }, cancellationToken);

        var csv = new StringBuilder();
        csv.AppendLine("Timestamp,User,Action,Entity,EntityId,Store,IpAddress,Details");

        foreach (var log in logs.Items)
        {
            csv.AppendLine($"\"{log.Timestamp:yyyy-MM-dd HH:mm:ss}\",\"{log.UserName}\",\"{log.Action}\",\"{log.EntityName}\",\"{log.EntityId}\",\"{log.StoreName}\",\"{log.IpAddress}\",\"{log.Details}\"");
        }

        return Encoding.UTF8.GetBytes(csv.ToString());
    }

    public async Task<byte[]> ExportSecurityLogsAsync(SecurityLogSearchRequest request, CancellationToken cancellationToken = default)
    {
        var logs = await GetSecurityLogsAsync(new SecurityLogSearchRequest
        {
            StartDate = request.StartDate,
            EndDate = request.EndDate,
            UserId = request.UserId,
            EventType = request.EventType,
            Severity = request.Severity,
            Success = request.Success,
            StoreId = request.StoreId,
            Page = 1,
            PageSize = 10000 // Large page for export
        }, cancellationToken);

        var csv = new StringBuilder();
        csv.AppendLine("Timestamp,EventType,Severity,User,Description,IpAddress,Success,Store");

        foreach (var log in logs.Items)
        {
            csv.AppendLine($"\"{log.Timestamp:yyyy-MM-dd HH:mm:ss}\",\"{log.EventTypeName}\",\"{log.SeverityName}\",\"{log.UserName}\",\"{log.Description}\",\"{log.IpAddress}\",\"{log.Success}\",\"{log.StoreName}\"");
        }

        return Encoding.UTF8.GetBytes(csv.ToString());
    }

    #endregion
}
